### DISCUSSION TITLE
Script for using fzf to select layout

### DISCUSSION THREAD

**Discussion by brandonchinn178** [kovidgoyal/kitty#5800] [Show and tell]
Couldn't find a solution in existing discussions, so wrote my own. Includes an asterisk by the currently active layout.

<img width="206" alt="Screen Shot 2022-12-20 at 12 41 20 PM" src="https://user-images.githubusercontent.com/6827922/208762349-1356cbac-dcd7-4caf-80b6-cf7c0393e31b.png">

```bash
#!/usr/bin/env sh

active_tab="$(
    kitty @ ls | jq -r '
        .[]
        | select(.is_active)
        | .tabs[]
        | select(.is_active_tab)
    '
)"

tab_id="$(jq -r '.id' <<< "${active_tab}")"
tab_layouts="$(jq -r '
    . as $tab
    | .enabled_layouts[]
    | (if . == $tab.layout then "*" else " " end) + " " + .
' <<< "${active_tab}")"

selected_layout="$(fzf --reverse <<< "${tab_layouts}" | cut -c 3-)"

kitty @ goto-layout -m "id:${tab_id}" "${selected_layout}"
```
```
env PATH=/usr/local/bin:/opt/homebrew/bin:$PATH

map cmd+l toggle_layout stack
map cmd+shift+l launch --allow-remote-control --type=overlay --title='*Layout*' /path/to/script
map ctrl+l next_layout
```

**Comment by page-down** [comment]
> ... launch only seems to add ...

Programs run via launch will use the environment variables seen by kitty. Since kitty is launched by system launchd, not the shell, there is no configured PATH from the shell.

https://sw.kovidgoyal.net/kitty/faq/#things-behave-differently-when-running-kitty-from-system-launcher-vs-from-another-terminal

```conf
map f1 show_kitty_env_vars
```

> ... Not sure if there's a better way

https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.env

```conf
env PATH=/usr/local/bin:$PATH
```

`exe_search_path` will not change the environment variable and will only be used to find the path to the program run directly by `launch`, and will not be helpful for the program in the script.
Under macOS, it is generally recommended to configure `exe_search_path`, because environment variables from launchd are tricky to configure.

https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.exe_search_path

```conf
exe_search_path +/usr/local/bin
```

Also, for `shebang`, this will not work if `zsh` is not in the PATH that kitty can see and PATH is not configured via `env`.
```text
#!/usr/bin/env zsh
```

Be careful not to start programs that need to work with TTY at the same time in a pipe.
In the last line, the alternative to `xargs` is to save the layout name selected by fzf first, and then call `kitty @`.

> Includes an asterisk by the currently active layout.

When the final selection is the currently active layout, then the script can exit without executing `kitty @`.

If no special zsh features are used in the script (after you have configured `env PATH`), then `sh` can be used to increase portability for use on different platforms that do not have zsh installed.

**Reply by brandonchinn178** [reply]
Thanks! The `env` option helps with `PATH` (sadly, though, I have to hardcode brew's symlink locations. I share configuration between an M2 macbook and an intel macbook, and brew has different locations for each: `/opt/homebrew` and `/usr/local`).

> Be careful not to start programs that need to work with TTY at the same time in a pipe.
> In the last line, the alternative to `xargs` is to save the layout name selected by fzf first, and then call `kitty @`.

So actually this script and your [previously linked script](https://github.com/kovidgoyal/kitty/issues/1303#issuecomment-1311202869) don't work with `sh` or `bash`. `sh` and `bash` strips newlines when interpolating stdout in `$(...)`. Is your comment here ("programs that need to work with TTY") a reference to fzf or kitty? I'm pretty sure `fzf` is intended to run in a pipe: https://github.com/junegunn/fzf/issues/1907#issuecomment-593996676

But I've updated the post to work with `sh`. I'll comment on your other issue to fix the script

**Reply by page-down** [reply]
> Is your comment here ("...") a reference to fzf or kitty?
> I'm pretty sure fzf is intended to run in a pipe ...

Normally there is no problem using pipes, however `kitty @` will require the TTY by default if you do not use socket for remote control.

```text
https://github.com/kovidgoyal/kitty/issues/5159#issuecomment-1202346340

> I doubt that would work ... If it does it would be very accidental.
> In general you cannot run multiple programs in a pipeline that expect to do terminal IO.
> Here both fzf and kitty @ are expecting to do terminal IO, simultaneously.
```

> So actually ... your ... script don't work with sh or bash.

Oh, I'm using bash 5.2 (released around September 2022) and it works fine.
Anyway, thanks for pointing that out. This will be useful for other users.

**Reply by brandonchinn178** [reply]
ah makes sense. I just use the bash shipped with Mac, which is 3.2
